docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,159 @@
+
+Once Gardener clusters are migrated to use CRI, we will implement two Gardener extensions, providing gVisor and Kata Containers as options for Gardener landscapes. The `WorkerGroup` specification will be extended to allow specifying a list of additional required Runtimes for nodes in that group. For example:
+```yaml
+workers:
+- name: worker-b8jg5
+  machineType: m5.large
+  volumeType: gp2
+  volumeSize: 50Gi
+  autoScalerMin: 1
+  autoScalerMax: 2
+  maxSurge: 1
+  additionalRuntimes:
+  - gvisor
+  - kata-containers
+  machineImage:
+    name: coreos
+    version: 2135.6.0
+```
+
+Each extension will need to address the following concern:
+
+1. Add the low-level runtime binaries to the worker nodes. Each extension should store the runtime binary in a way that is accessible to a node to download and install during startup.
+1. Hook the runtime binary into the containerd configuration file, so that the runtime becomes available to containerd.
+1. Apply a label to each node that allows identifying nodes where the runtime is available.
+1. Apply the relevant `RuntimeClass` to the Gardener cluster, to expose the functionality to users.
+1. Provide a `ValidatingWebhook` to catch invalid configurations. For example, Kata Containers on AWS requires a `machineType` of `i3.metal`.
+
+Since each operating system distribution has different methods of installing software (apt on ubuntu, zypper on SuSE, Torcx on CoreOS/Flatcar), we will enhance the operating system extensions to add a standard `installSoftware` script which can abstract the mechanics of installing new software on a node. The new runtime extensions will use this new functionality to install the relevant runtimes to the worker nodes.
+
+## Design Details
+
+1. In order to configure each worker machine in the cluster to work with CRI, the following configurations should be done:
+    1. Add kubelet execution flags:
+        1. --container-runtime=remote
+        2. --container-runtime-endpoint=unix:///.../containerd.socket
+    2. Make sure that containerd configuration file exist in path /etc/containerd/config.toml and it is configured to enable the CRI plugin (In the containerd config.toml "cri" plugin must not be disabled).
+    
+2. Containerd socket path and containerd configurations are different for each OS. To make sure the configurations above are set in each worker machine, each OS extension would be responsible to configure them.
+    1. os-ubuntu - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
```

Comment:
Hm, do we really need the webhook here? Couldn't the OSC controllers simply add the `--container-runtime-endpoint` flag as part of their computation of the user-data/cloud-init?

Reply:
You are right, we can do it as part of the OSC reconciliation flow, i will update it.